lib/Runtime/Library/TypedArray.cpp

```diff
@@ -143,7 +143,11 @@ namespace Js
             {
                 if (JavascriptOperators::IsObject(firstArgument))
                 {
-                    if (JavascriptOperators::IsIterable(RecyclableObject::FromVar(firstArgument), scriptContext))
+                    if (JavascriptOperators::IsIterable(RecyclableObject::FromVar(firstArgument), scriptContext) &&
+                        (!JavascriptArray::Is(firstArgument) ||
+                        JavascriptOperators::HasOwnProperty(firstArgument, PropertyIds::_symbolIterator, scriptContext) ||
+                        DynamicObject::FromVar(firstArgument)->GetPrototype() != scriptContext->GetLibrary()->GetArrayPrototype() ||
+                        scriptContext->GetLibrary()->GetArrayObjectHasUserDefinedIterator()))
```

Comment:
This is unnecessarily complicated.

`IsIterable` is also suspicious.  To me it looks like for proxy objects we are going to call the `Symbol.iterator` property on proxy objects twice instead of once which have observable side effects that don't match the spec semantics.

We should do exactly as the spec says.  Get the `Symbol.iterator` property's value from the object.  At that point, before calling it, test to see if it is the built-in array iterator function.  If it is, _then_ check if the object to see if it is an array and do the fast path.  Otherwise continue with the rest of the spec algorithm.

Note that if the the array object did have its own `Symbol.iterator` but it was the built-in array iterator method we could still do the optimization, so there is no need to check only the prototype case.  There is no need to handle the cases separately at all since we always have to do a `Get` of `Symbol.iterator` on the object which will do the prototype lookup.  It is only the value returned from this `Get` that matters.